### PR TITLE
Stream sync protocol

### PR DESCRIPTION
**Pull request by JackyWYX**
## Important notice

Since the size of this PR is too large for review, will split this PR to a set of small PRs for review. The breakdown PR will follow a bottom-up sequence for review.

1. [Fix p2p base stream read/write](https://github.com/harmony-one/harmony/pull/3552)
2. [Stream manager](https://github.com/harmony-one/harmony/pull/3554)
3. [Rate limiter](https://github.com/harmony-one/harmony/pull/3553)
4. [Protobuf message](https://github.com/harmony-one/harmony/pull/3561)
5. [Request manager](https://github.com/harmony-one/harmony/pull/3560)

## Description

1. Added stream sync protocol running on P2P Stream.
2. Added downloader module to manage the synchronization task.
3. Stream manager module to manage over streams.
4. Request manager module to manage over stream requests.
5. Flags, commands and misc.

## Tests

1. 1-1 sync pair test - Done
2. 1-3 sync test - Done
3. 1-3 with 1 bad peer, 2 good peer (negative test)  (Almost finished)
4. Large scale test (STN test) - Not started.

## Current known issues

1. In very rare cases got error when establishing stream connection `failed to negotiate security protocol: error reading handshake message: noise: message is too short` (https://github.com/libp2p/go-libp2p-noise/issues/70) This issue is already fixed in libp2p version v0.13.0. Will upgrade libp2p after this release (frozen code).

### CODE REVIEW

File: cmd/harmony/config.go [hunk 1]
```diff
@@ -151,6 +153,18 @@ type prometheusConfig struct {
 	Gateway    string
 }
 
+type syncConfig struct {
+	LegacyServer   bool
```
Comment (rlan35) [inline cmd/harmony/config.go:157]: can we add some comments on what these fields mean?

File: cmd/harmony/main.go [hunk 2]
```diff
@@ -375,8 +371,15 @@ func setupNodeAndRun(hc harmonyConfig) {
 		setupPrometheusService(currentNode, hc, nodeConfig.ShardID)
 	}
 
-	// TODO: replace this legacy syncing
-	currentNode.SupportSyncing()
+	// TODO: probably replace this legacy syncing after a hard fork
```
Comment (rlan35) [inline cmd/harmony/main.go:374]: why need a hard fork?

Comment (JackyWYX) [reply cmd/harmony/main.go:374]: Nah, we don't need it. Removed comment.

File: hmy/downloader/downloaders.go [hunk 3]
```diff
@@ -0,0 +1,54 @@
+package downloader
+
+import (
+	"github.com/harmony-one/harmony/core"
+	"github.com/harmony-one/harmony/p2p"
+)
+
+// Downloaders is the set of downloaders
+type Downloaders struct {
+	ds map[uint32]*Downloader
```
Comment (rlan35) [inline hmy/downloader/downloaders.go:10]: is there race condition on map access?

Comment (JackyWYX) [reply hmy/downloader/downloaders.go:10]: No. Currently, this map is fixed after setup. We will need to add mutex probably for resharding. For now, I think it's good.

File: cmd/harmony/config.go (hunk 1 above)
Comment (JackyWYX) [reply cmd/harmony/config.go]: Sure. Added.

**Comment by rlan35** [comment]
Are we ready to test this on a real network?

**Comment by JackyWYX** [comment]
> Are we ready to test this on a real network?

Yep, ready! :) @LeoHChen

File: consensus/consensus_v2.go [hunk 4]
```diff
@@ -467,13 +467,25 @@ func (consensus *Consensus) Start(
 		}
 		consensus.getLogger().Info().Msg("[ConsensusMainLoop] Ended.")
 	}()
+
+	if consensus.dHelper != nil {
+		consensus.dHelper.start()
+	}
```
Comment (LeoHChen) [inline consensus/consensus_v2.go:473]: any error message if dHelper == nil?

File: consensus/consensus_v2.go [hunk 5]
```diff
@@ -467,13 +467,25 @@ func (consensus *Consensus) Start(
...
 // phase end.
 func (consensus *Consensus) Close() error {
+	if consensus.dHelper != nil {
+		consensus.dHelper.close()
+	}
+	consensus.waitForCommit()
+	return nil
+}
+
+func (consensus *Consensus) waitForCommit() {
```
Comment (LeoHChen) [inline consensus/consensus_v2.go:486]: add a comment to this function please.

**Review by LeoHChen — COMMENTED** [review]
some quick comments for this huge PR. Only go through a few files.

**Comment by JackyWYX** [comment]
Since most of the code in this branch is already done in https://github.com/harmony-one/harmony/pull/3588, closing!

**Pull request closed without merging**